README.md
# Aphrodite [![npm version](https://badge.fury.io/js/aphrodite.svg)](https://badge.fury.io/js/aphrodite) [![Build Status](https://travis-ci.org/Khan/aphrodite.svg?branch=master)](https://travis-ci.org/Khan/aphrodite) [![Coverage Status](https://coveralls.io/repos/github/Khan/aphrodite/badge.svg?branch=master)](https://coveralls.io/github/Khan/aphrodite?branch=master) [![Gitter chat](https://img.shields.io/gitter/room/Khan/aphrodite.svg)](https://gitter.im/Khan/aphrodite) [![gzip size][gzip-badge]][unpkg-dist] [![size][size-badge]][unpkg-dist]

_Framework-agnostic CSS-in-JS with support for server-side rendering, browser prefixing, and minimum CSS generation._

Support for colocating your styles with your JavaScript component.

- Works great with and without React
- Supports media queries without window.matchMedia
- Supports pseudo-selectors like `:hover`, `:active`, etc. without needing to
  store hover or active state in components. `:visited` works just fine too.
- Supports automatic global `@font-face` detection and insertion.
- Respects precedence order when specifying multiple styles
- Requires no AST transform
- Injects only the exact styles needed for the render into the DOM.
- Can be used for server rendering
- Few dependencies, small (20k, 6k gzipped)
- No external CSS file generated for inclusion
- Autoprefixes styles

# Installation

Aphrodite is distributed via [npm](https://www.npmjs.com/):

```
npm install --save aphrodite
```

# API

If you'd rather watch introductory videos, you can find them [here](https://www.youtube.com/playlist?list=PLo4Zh55ZzNSBP78pCD0dZJi9zf8CA72_M).

```jsx
import React, { Component } from 'react';
import { StyleSheet, css } from 'aphrodite';

class App extends Component {
    render() {
        return <div>
            <span className={css(styles.red)}>
                This is red.
            </span>
            <span className={css(styles.hover)}>
                This turns red on hover.
            </span>
            <span className={css(styles.small)}>
                This turns red when the browser is less than 600px width.
            </span>
            <span className={css(styles.red, styles.blue)}>
                This is blue.
            </span>
            <span className={css(styles.blue, styles.small)}>
                This is blue and turns red when the browser is less than
                600px width.
            </span>
        </div>;
    }
}

const styles = StyleSheet.create({
    red: {
        backgroundColor: 'red'
    },

    blue: {
        backgroundColor: 'blue'
    },

    hover: {
        ':hover': {
            backgroundColor: 'red'
        }
    },

    small: {
        '@media (max-width: 600px)': {
            backgroundColor: 'red',
        }
    }
});
```

## Conditionally Applying Styles

Note: If you want to conditionally use styles, that is simply accomplished via:

```jsx
const className = css(
  shouldBeRed() ? styles.red : styles.blue,
  shouldBeResponsive() && styles.small,
  shouldBeHoverable() && styles.hover
)

<div className={className}>Hi</div>
```

This is possible because any falsey arguments will be ignored.

## Combining Styles

To combine styles, pass multiple styles or arrays of styles into `css()`. This is common when combining styles from an owner component:

```jsx
class App extends Component {
    render() {
        return <Marker styles={[styles.large, styles.red]} />;
    }
}

class Marker extends Component {
    render() {
        // css() accepts styles, arrays of styles (including nested arrays),
        // and falsy values including undefined.
        return <div className={css(styles.marker, this.props.styles)} />;
    }
}

const styles = StyleSheet.create({
    red: {
        backgroundColor: 'red'
    },

    large: {
        height: 20,
        width: 20
    },

    marker: {
        backgroundColor: 'blue'
    }
});
```

## Resetting Style Cache

The `reset` function can be used to reset the HTML style tag, injection buffer, and injected cache. Useful when Aphrodite needs to be torn down and set back up.

```js
import { reset } from 'aphrodite';

reset();
```

While the `resetInjectedStyle` function can be used to reset the injected cache for a single key (usually the class name).

```js
import { resetInjectedStyle } from 'aphrodite';

resetInjectedStyle('class_1sAs8jg');
```

## Server-side rendering

To perform server-side rendering, make a call to `StyleSheetServer.renderStatic`, which takes a callback. Do your rendering inside of the callback and return the generated HTML. All of the calls to `css()` inside of the callback will be collected and the generated css as well as the generated HTML will be returned.

Rehydrating lets Aphrodite know which styles have already been inserted into the page. If you don't rehydrate, Aphrodite might add duplicate styles to the page.

To perform rehydration, call `StyleSheet.rehydrate` with the list of generated class names returned to you by `StyleSheetServer.renderStatic`.

Note: If you are using `aphrodite/no-important` in your project and you want to render it on server side, be sure to import `StyleSheetServer` from `aphrodite/no-important` otherwise you are going to get an error.

As an example:

```js
import { StyleSheetServer } from 'aphrodite';

// Contains the generated html, as well as the generated css and some
// rehydration data.
var {html, css} = StyleSheetServer.renderStatic(() => {
    return ReactDOMServer.renderToString(<App/>);
});

// Return the base HTML, which contains your rendered HTML as well as a
// simple rehydration script.
return `
    <html>
        <head>
            <style data-aphrodite>${css.content}</style>
        </head>
        <body>
            <div id='root'>${html}</div>
            <script src="./bundle.js"></script>
            <script>
                StyleSheet.rehydrate(${JSON.stringify(css.renderedClassNames)});
                ReactDOM.render(<App/>, document.getElementById('root'));
            </script>
        </body>
    </html>
`;
```

## Disabling `!important`

By default, Aphrodite will append `!important` to style definitions. This is
intended to make integrating with a pre-existing codebase easier. If you'd like
to avoid this behaviour, then instead of importing `aphrodite`, import
`aphrodite/no-important`. Otherwise, usage is the same:

```js
import { StyleSheet, css } from 'aphrodite/no-important';
```

## Minifying style names

By default, Aphrodite will minify style names down to their hashes in production
(`process.env.NODE_ENV === 'production'`). You can override this behavior by
calling `minify` with `true` or `false` before calling `StyleSheet.create`.

This is useful if you want to facilitate debugging in production for example.

```js
import { StyleSheet, minify } from 'aphrodite';

// Always keep the full style names
minify(false);

// ... proceed to use StyleSheet.create etc.
```

## Font Faces

Creating custom font faces is a special case. Typically you need to define a global `@font-face` rule. In the case of Aphrodite we only want to insert that rule if it's actually being referenced by a class that's in the page. We've made it so that the `fontFamily` property can accept a font-face object (either directly or inside an array). A global `@font-face` rule is then generated based on the font definition.

```js
const coolFont = {
    fontFamily: "CoolFont",
    fontStyle: "normal",
    fontWeight: "normal",
    src: "url('coolfont.woff2') format('woff2')"
};

const styles = StyleSheet.create({
    headingText: {
        fontFamily: coolFont,
        fontSize: 20
    },
    bodyText: {
        fontFamily: [coolFont, "sans-serif"]
        fontSize: 12
    }
});
```

Aphrodite will ensure that the global `@font-face` rule for this font is only inserted once, no matter how many times it's referenced.

## Animations

Similar to [Font Faces](#font-faces), Aphrodite supports keyframe animations, but it's treated as a special case. Once we find an instance of the animation being referenced, a global `@keyframes` rule is created and appended to the page.

Animations are provided as objects describing the animation, in typical `@keyframes` fashion. Using the `animationName` property, you can supply a single animation object, or an array of animation objects. Other animation properties like `animationDuration` can be provided as strings.

```js
const translateKeyframes = {
    '0%': {
        transform: 'translateX(0)',
    },

    '50%': {
        transform: 'translateX(100px)',
    },

    '100%': {
        transform: 'translateX(0)',
    },
};

const opacityKeyframes = {
    'from': {
        opacity: 0,
    },

    'to': {
        opacity: 1,
    }
};

const styles = StyleSheet.create({
    zippyHeader: {
        animationName: [translateKeyframes, opacityKeyframes],
        animationDuration: '3s, 1200ms',
        animationIterationCount: 'infinite',
    },
});
```

Aphrodite will ensure that `@keyframes` rules are never duplicated, no matter how many times a given rule is referenced.

# Use without React

Aphrodite was built with React in mind but does not depend on React. Here, you can see it
used with [Web Components][webcomponents]:

```js
import { StyleSheet, css } from 'aphrodite';

const styles = StyleSheet.create({
    red: {
        backgroundColor: 'red'
    }
});

class App extends HTMLElement {
    attachedCallback() {
        this.innerHTML = `
            <div class="${css(styles.red)}">
                This is red.
            </div>
        `;
    }
}

document.registerElement('my-app', App);
```

# Caveats

## Style injection and buffering

Aphrodite will automatically attempt to create a `<style>` tag in the document's `<head>` element to put its generated styles in. Aphrodite will only generate one `<style>` tag and will add new styles to this over time. If you want to control which style tag Aphrodite uses, create a style tag yourself with the `data-aphrodite` attribute and Aphrodite will use that instead of creating one for you.

To speed up injection of styles, Aphrodite will automatically try to buffer writes to this `<style>` tag so that minimum number of DOM modifications happen.

Aphrodite uses [asap](https://github.com/kriskowal/asap) to schedule buffer flushing. If you measure DOM elements' dimensions in `componentDidMount` or `componentDidUpdate`, you can use `setTimeout` or `flushToStyleTag` to ensure all styles are injected.

```js
import { StyleSheet, css } from 'aphrodite';

class Component extends React.Component {
    render() {
        return <div ref="root" className={css(styles.div)} />;
    }

    componentDidMount() {
        // At this point styles might not be injected yet.
        this.refs.root.offsetHeight; // 0 or 10

        setTimeout(() => {
            this.refs.root.offsetHeight; // 10
        }, 0);
    }
}

const styles = StyleSheet.create({
    div: {
        height: 10,
    },
});
```

## Assigning a string to a content property for a pseudo-element

When assigning a string to the `content` property it requires double or single quotes in CSS.
Therefore with Aphrodite you also have to provide the quotes within the value string for `content` to match how it will be represented in CSS.

As an example:

```javascript
const styles = StyleSheet.create({
  large: {
      ':after': {
        content: '"Aphrodite"',
      },
    },
  },
  small: {
      ':before': {
        content: "'Aphrodite'",
      },
    },
  });
```
The generated css will be:

```css
  .large_im3wl1:after {
      content: "Aphrodite" !important;
  }

  .small_ffd5jf:before {
      content: 'Aphrodite' !important;
  }
```

## Overriding styles

When combining multiple aphrodite styles, you are strongly recommended to merge all of your styles into a single call to `css()`, and should not combine the generated class names that aphrodite outputs (via string concatenation, `classnames`, etc.).
For example, if you have a base style of `foo` which you are trying to override with `bar`:

### Do this:

```js
const styles = StyleSheet.create({
  foo: {
    color: 'red'
  },

  bar: {
    color: 'blue'
  }
});

// ...

const className = css(styles.foo, styles.bar);
```

### Don't do this:

```js
const styles = StyleSheet.create({
  foo: {
    color: 'red'
  },

  bar: {
    color: 'blue'
  }
});

// ...

const className = css(styles.foo) + " " + css(styles.bar);
```

Why does it matter? Although the second one will produce a valid class name, it cannot guarantee that the `bar` styles will override the `foo` ones.
The way the CSS works, it is not the *class name that comes last on an element* that matters, it is specificity. When we look at the generated CSS though, we find that all of the class names have the same specificity, since they are all a single class name:

```css
.foo_im3wl1 {
  color: red;
}
```

```css
.bar_hxfs3d {
  color: blue;
}
```

In the case where the specificity is the same, what matters is *the order that the styles appear in the stylesheet*. That is, if the generated stylesheet looks like

```css
.foo_im3wl1 {
  color: red;
}
.bar_hxfs3d {
  color: blue;
}
```

then you will get the appropriate effect of the `bar` styles overriding the `foo` ones, but if the stylesheet looks like

```css
.bar_hxfs3d {
  color: blue;
}
.foo_im3wl1 {
  color: red;
}
```

then we end up with the opposite effect, with `foo` overriding `bar`! The way to solve this is to pass both of the styles into aphrodite's `css()` call. Then, it will produce a single class name, like `foo_im3wl1-o_O-bar_hxfs3d`, with the correctly overridden styles, thus solving the problem:

```css
.foo_im3wl1-o_O-bar_hxfs3d {
  color: blue;
}
```

## Object key ordering

When styles are specified in Aphrodite, the order that they appear in the
actual stylesheet depends on the order that keys are retrieved from the
objects. This ordering is determined by the JavaScript engine that is being
used to render the styles. Sometimes, the order that the styles appear in the
stylesheet matter for the semantics of the CSS. For instance, depending on the
engine, the styles generated from

```js
const styles = StyleSheet.create({
    ordered: {
        margin: 0,
        marginLeft: 15,
    },
});
css(styles.ordered);
```

you might expect the following CSS to be generated:

```css
margin: 0px;
margin-left: 15px;
```

but depending on the ordering of the keys in the style object, the CSS might
appear as

```css
margin-left: 15px;
margin: 0px;
```

which is semantically different, because the style which appears later will
override the style before it.

This might also manifest as a problem when server-side rendering, if the
generated styles appear in a different order on the client and on the server.

If you experience this issue where styles don't appear in the generated CSS in
the order that they appear in your objects, there are two solutions:

1. Don't use shorthand properties. For instance, in the margin example above,
   by switching from using a shorthand property and a longhand property in the
   same styles to using only longhand properties, the issue could be avoided.

   ```js
   const styles = StyleSheet.create({
       ordered: {
           marginTop: 0,
           marginRight: 0,
           marginBottom: 0,
           marginLeft: 15,
       },
   });
   ```

2. Specify the ordering of your styles by specifying them using a
   [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map).
   Since `Map`s preserve their insertion order, Aphrodite is able to place your
   styles in the correct order.

   ```js
   const styles = StyleSheet.create({
       ordered: new Map([
           ["margin", 0],
           ["marginLeft", 15],
       ]),
   });
   ```

   Note that `Map`s are not fully supported in all browsers. It can be
   polyfilled by using a package
   like [es6-shim](https://www.npmjs.com/package/es6-shim).

## Advanced: Extensions

Extra features can be added to Aphrodite using extensions.

To add extensions to Aphrodite, call `StyleSheet.extend` with the extensions
you are adding. The result will be an object containing the usual exports of
Aphrodite (`css`, `StyleSheet`, etc.) which will have your extensions included.
For example:

```js
// my-aphrodite.js
import {StyleSheet} from "aphrodite";

export default StyleSheet.extend([extension1, extension2]);

// styled.js
import {StyleSheet, css} from "my-aphrodite.js";

const styles = StyleSheet.create({
    ...
});
```

**Note**: Using extensions may cause Aphrodite's styles to not work properly.
Plain Aphrodite, when used properly, ensures that the correct styles will
always be applied to elements. Due to CSS specificity rules, extensions might
allow you to generate styles that conflict with each other, causing incorrect
styles to be shown. See the global extension below to see what could go wrong.

### Creating extensions

Currently, there is only one kind of extension available: selector handlers.
These kinds of extensions let you look at the selectors that someone specifies
and generate new selectors based on them. They are used to handle pseudo-styles
and media queries inside of Aphrodite. See the
[`defaultSelectorHandlers` docs](src/generate.js?L8) for information about how
to create a selector handler function.

To use your extension, create an object containing a key of the kind of
extension that you created, and pass that into `StyleSheet.extend()`:

```js
const mySelectorHandler = ...;

const myExtension = {selectorHandler: mySelectorHandler};

const { StyleSheet: newStyleSheet, css: newCss } = StyleSheet.extend([myExtension]);
```

As an example, you could write an extension which generates global styles like

```js
const globalSelectorHandler = (selector, _, generateSubtreeStyles) => {
    if (selector[0] !== "*") {
        return null;
    }

    return generateSubtreeStyles(selector.slice(1));
};

const globalExtension = {selectorHandler: globalSelectorHandler};
```

This might cause problems when two places try to generate styles for the same
global selector however! For example, after

```js

CHANGELOG.md
# 2.4.0

- Allow already injected to be reset ([#380](https://github.com/Khan/aphrodite/pull/380))
- Add TypeScript export for flushToStyleTag ([#379](https://github.com/Khan/aphrodite/pull/379))
- Make parameter typings for StyleSheet.create more typed ([#375](https://github.com/Khan/aphrodite/pull/375))
- Add a more explicit error message when someone tries to pass a plain object instead of a stylesheet into the css() method ([#349](https://github.com/Khan/aphrodite/pull/349))
- Roll up inline-style-prefixer ([#364](https://github.com/Khan/aphrodite/pull/364))
- Update inline-style-prefixer 5.0.4 → 5.1.0 ([#373](https://github.com/Khan/aphrodite/pull/373))
- Fix examples ([#367](https://github.com/Khan/aphrodite/pull/367))

# 2.3.1

- Use deep import for inline-style-prefixer ([#363](https://github.com/Khan/aphrodite/pull/363))

# 2.3.0

- Bump inline-style-prefixer 4.0.2 -> 5.0.4 ([#358](https://github.com/Khan/aphrodite/pull/358))
- Update rollup to v1.2.0 and babel to v7 ([#362](https://github.com/Khan/aphrodite/pull/362))

# 2.2.3

- Bump inline-style-prefixer 4.0.0 -> 4.0.2 ([#341](https://github.com/Khan/aphrodite/pull/341))

# 2.2.2

- Fix multiple `@font-face` of the same name ([#327](https://github.com/Khan/aphrodite/pull/327))
- Fix typings to allow for extending ([#330](https://github.com/Khan/aphrodite/pull/330))

# 2.2.1

- Use code-splitting for default and no-important builds ([#325](https://github.com/Khan/aphrodite/pull/325))

# 2.2.0

- Expose `defaultSelectorHandlers` and `injectAndGetClassName` ([#320](https://github.com/Khan/aphrodite/pull/320))


# 2.1.1

- Add `minify` and `flushToStyleTag` to no-important ([#316](https://github.com/Khan/aphrodite/pull/316))

# 2.1.0

- Expose `flushToStyleTag` ([#310](https://github.com/Khan/aphrodite/pull/310))

# 2.0.0

- Selector handlers now must return an array of rules. Returning a string containing multiple rules is deprecated. The fallback behavior for selector handlers that still return strings is to use media queries, which are not supported in older browsers like Internet Explorer 8, which makes this a breaking change. If you find yourself seeing this warning, you need to update your custom selector handlers to return arrays of strings instead of plain strings. This fallback will be removed entirely in a future major release.
- `insertRule` is now used to inject styles to the style element ([#240](https://github.com/Khan/aphrodite/pull/240))
- New `minify()` controls the minification of style names (defaults to `true` when `process.env.NODE_ENV === 'production'`) ([#291](https://github.com/Khan/aphrodite/pull/291))
- New TypeScript typings ([#302](https://github.com/Khan/aphrodite/pull/302))
- New `StyleSheetTestUtils.getBufferedStyles()` ([#294](https://github.com/Khan/aphrodite/pull/294))
- `StyleSheetTestUtils` will be minified out in production builds (i.e. when `process.env.NODE_ENV === 'production'`) ([#305](https://github.com/Khan/aphrodite/pull/305))
- Reduce bundle size impact by building with rollup instead of webpack ([#281](https://github.com/Khan/aphrodite/pull/281))
- Simplify output for more useful RunKit result ([#244](https://github.com/Khan/aphrodite/pull/244))
- Update inline-style-prefixer dependency to v4 ([#297](https://github.com/Khan/aphrodite/pull/297))

# 1.2.5

- Allow overriding styles to re-order the styles they override ([#279](https://github.com/Khan/aphrodite/pull/279))

# 1.2.4

- Minify combined class names in production builds ([#248](https://github.com/Khan/aphrodite/pull/248))

# 1.2.3

- Revert make string handlers use `useImportant`([f162220e3](https://github.com/Khan/aphrodite/commit/f162220e3d7321c7ede0aefc189de9bb694e5107)

# 1.2.2

- Minify selectors when building with `process.env.NODE_ENV === 'production'` ([#246](https://github.com/Khan/aphrodite/pull/246))
- Make string handlers use `useImportant` ([#256](https://github.com/Khan/aphrodite/pull/256))

# 1.2.1

- Fix bug that caused unexpected mutations of nested objects ([#231](https://github.com/Khan/aphrodite/issues/231))
- More optimizations ([#233](https://github.com/Khan/aphrodite/pull/233), [#216](https://github.com/Khan/aphrodite/pull/216))

# 1.2.0

- Allow specifying styles as `Map`s to guarantee ordering ([#200](https://github.com/Khan/aphrodite/pull/200))
- Replace murmur hash with djb2 hash. This changes all generated classNames. ([#203](https://github.com/Khan/aphrodite/pull/203))
- Misc. optimizations ([#201](https://github.com/Khan/aphrodite/pull/201), [#202](https://github.com/Khan/aphrodite/pull/202), [#204](https://github.com/Khan/aphrodite/pull/204), [#205](https://github.com/Khan/aphrodite/pull/205), [#206](https://github.com/Khan/aphrodite/pull/206), [#207](https://github.com/Khan/aphrodite/pull/207), [#208](https://github.com/Khan/aphrodite/pull/208), [#213](https://github.com/Khan/aphrodite/pull/213))

# 1.1.0

- Animations now support multiple animations per style ([see section on Animations](https://github.com/Khan/aphrodite#animations)) ([PR #167](https://github.com/Khan/aphrodite/pull/167))

# 1.0.0
- Syntax extensions ([see section on Advanced extensions](https://github.com/Khan/aphrodite#advanced-extensions)) ([PR #95](https://github.com/Khan/aphrodite/pull/95))

# 0.6.0
- `css()` will now accept arbitrarily nested arrays. i.e. instead of `css(styles.a, styles.b)`, you can now do `css([styles.a, [styles.b, styles.c]])`. ([PR #154](https://github.com/Khan/aphrodite/pull/154))
- Support for multiple font styles with the same font-family. ([PR #82](https://github.com/Khan/aphrodite/pull/82))


src/inject.js
/* @flow */
import asap from 'asap';

import OrderedElements from './ordered-elements';
import { generateCSS } from './generate';
import { hashObject, hashString } from './util';

/* ::
import type { SheetDefinition, SheetDefinitions } from './index.js';
import type { MaybeSheetDefinition } from './exports.js';
import type { SelectorHandler } from './generate.js';
*/

export class Inject {
    constructor() {

        // The current <style> tag we are inserting into, or null if we haven't
        // inserted anything yet. We could find this each time using
        // `document.querySelector("style[data-aphrodite"])`, but holding onto it is
        // faster.
        /** @type {any} */
        this.styleTag = null;
        // This is a map from Aphrodite's generated class names to `true` (acting as a
        // set of class names)
        this.alreadyInjected = {};

        // This is the buffer of styles which have not yet been flushed.
        /** @type {string[]} */
        this.injectionBuffer = [];

        // A flag to tell if we are already buffering styles. This could happen either
        // because we scheduled a flush call already, so newly added styles will
        // already be flushed, or because we are statically buffering on the server.
        this.isBuffering = false;

        this.stringHandlers = { animationName: this.animationName.bind(this), fontFamily: this.fontFamily.bind(this) };

        this.flushToStyleTag = () => {
            const cssRules = this.flushToArray();
            if (cssRules.length > 0) {
                this.injectStyleTag(cssRules);
            }
        }
    }

    // Inject a set of rules into a <style> tag in the head of the document. This
    // will automatically create a style tag and then continue to use it for
    // multiple injections. It will also use a style tag with the `data-aphrodite`
    // tag on it if that exists in the DOM. This could be used for e.g. reusing the
    // same style tag that server-side rendering inserts.
    injectStyleTag(cssRules /* : string[] */) {
        if (this.styleTag == null) {
            // Try to find a style tag with the `data-aphrodite` attribute first.
            this.styleTag = ((document.querySelector("style[data-aphrodite]") /* : any */) /* : ?HTMLStyleElement */);

            // If that doesn't work, generate a new style tag.
            if (this.styleTag == null) {
                // Taken from
                // http://stackoverflow.com/questions/524696/how-to-create-a-style-tag-with-javascript
                const head = document.head || document.getElementsByTagName('head')[0];
                this.styleTag = document.createElement('style');

                this.styleTag.type = 'text/css';
                this.styleTag.setAttribute("data-aphrodite", "");
                head.appendChild(this.styleTag);
            }
        }

        // $FlowFixMe
        const sheet = ((this.styleTag.styleSheet || this.styleTag.sheet /* : any */) /* : CSSStyleSheet */);

        if (sheet.insertRule) {
            let numRules = sheet.cssRules.length;
            cssRules.forEach((rule) => {
                try {
                    sheet.insertRule(rule, numRules);
                    numRules += 1;
                } catch (e) {
                    // The selector for this rule wasn't compatible with the browser
                }
            });
        } else {
            this.styleTag.innerText = (this.styleTag.innerText || '') + cssRules.join('');
        }
    }

    // Custom handlers for stringifying CSS values that have side effects
    // (such as fontFamily, which can cause @font-face rules to be injected)
    // With fontFamily we look for objects that are passed in and interpret
    // them as @font-face rules that we need to inject. The value of fontFamily
    // can either be a string (as normal), an object (a single font face), or
    // an array of objects and strings.
    fontFamily(val) {
        if (Array.isArray(val)) {
            const nameMap = {};

            val.forEach(v => {
                nameMap[this.fontFamily(v)] = true;
            });

            return Object.keys(nameMap).join(",");
        } else if (typeof val === "object") {
            this.injectStyleOnce(val.src, "@font-face", [val], false);
            return `"${val.fontFamily}"`;
        } else {
            return val;
        }
    }

    // With animationName we look for an object that contains keyframes and
    // inject them as an `@keyframes` block, returning a uniquely generated
    // name. The keyframes object should look like
    //  animationName: {
    //    from: {
    //      left: 0,
    //      top: 0,
    //    },
    //    '50%': {
    //      left: 15,
    //      top: 5,
    //    },
    //    to: {
    //      left: 20,
    //      top: 20,
    //    }
    //  }
    // TODO(emily): `stringHandlers` doesn't let us rename the key, so I have
    // to use `animationName` here. Improve that so we can call this
    // `animation` instead of `animationName`.
    animationName(val, selectorHandlers) {
        if (Array.isArray(val)) {
            return val.map(v => this.animationName(v, selectorHandlers)).join(",");
        } else if (typeof val === "object") {
            // Generate a unique name based on the hash of the object. We can't
            // just use the hash because the name can't start with a number.
            // TODO(emily): this probably makes debugging hard, allow a custom
            // name?
            const name = `keyframe_${hashObject(val)}`;

            // Since keyframes need 3 layers of nesting, we use `generateCSS` to
            // build the inner layers and wrap it in `@keyframes` ourselves.
            let finalVal = `@keyframes ${name}{`;

            // TODO see if we can find a way where checking for OrderedElements
            // here is not necessary. Alternatively, perhaps we should have a
            // utility method that can iterate over either a plain object, an
            // instance of OrderedElements, or a Map, and then use that here and
            // elsewhere.
            if (val instanceof OrderedElements) {
                val.forEach((valVal, valKey) => {
                    finalVal += generateCSS(
                        valKey, [valVal], selectorHandlers, this.stringHandlers, false).join('');
                });
            } else {
                Object.keys(val).forEach(key => {
                    finalVal += generateCSS(
                        key, [val[key]], selectorHandlers, this.stringHandlers, false).join('');
                });
            }
            finalVal += '}';

            this.injectGeneratedCSSOnce(name, [finalVal]);

            return name;
        } else {
            return val;
        }
    }

    injectGeneratedCSSOnce(key, generatedCSS) {
        if (this.alreadyInjected[key]) {
            return;
        }

        if (!this.isBuffering) {
            // We should never be automatically buffering on the server (or any
            // place without a document), so guard against that.
            if (typeof document === "undefined") {
                throw new Error(
                    "Cannot automatically buffer without a document");
            }

            // If we're not already buffering, schedule a call to flush the
            // current styles.
            this.isBuffering = true;
            asap(this.flushToStyleTag);
        }

        this.injectionBuffer.push(...generatedCSS);
        this.alreadyInjected[key] = true;
    }

    injectStyleOnce(
        key /* : string */,
        selector /* : string */,
        definitions /* : SheetDefinition[] */,
        useImportant /* : boolean */,
        selectorHandlers /* : SelectorHandler[] */ = []
    ) {
        if (this.alreadyInjected[key]) {
            return;
        }

        const generated = generateCSS(
            selector, definitions, selectorHandlers,
            this.stringHandlers, useImportant);

        this.injectGeneratedCSSOnce(key, generated);
    }

    reset() {
        this.injectionBuffer = [];
        this.alreadyInjected = {};
        this.isBuffering = false;
        this.styleTag = null;
    }

    resetInjectedStyle(key /* : string */) {
        delete this.alreadyInjected[key];
    }

    getBufferedStyles() {
        return this.injectionBuffer;
    }

    startBuffering() {
        if (this.isBuffering) {
            throw new Error(
                "Cannot buffer while already buffering");
        }
        this.isBuffering = true;
    }

    flushToArray() {
        this.isBuffering = false;
        const ret = this.injectionBuffer;
        this.injectionBuffer = [];
        return ret;
    }

    flushToString() {
        return this.flushToArray().join('');
    }



    getRenderedClassNames() /* : string[] */ {
        return Object.keys(this.alreadyInjected);
    }

    addRenderedClassNames(classNames /* : string[] */) {
        classNames.forEach(className => {
            this.alreadyInjected[className] = true;
        });
    }

    isValidStyleDefinition(def /* : Object */) {
        return "_definition" in def && "_name" in def && "_len" in def;
    }

    processStyleDefinitions(
        styleDefinitions /* : any[] */,
        classNameBits /* : string[] */,
        definitionBits /* : Object[] */,
        length /* : number */,
    ) /* : number */ {
        for (let i = 0; i < styleDefinitions.length; i += 1) {
            // Filter out falsy values from the input, to allow for
            // `css(a, test && c)`
            if (styleDefinitions[i]) {
                if (Array.isArray(styleDefinitions[i])) {
                    // We've encountered an array, so let's recurse
                    length += this.processStyleDefinitions(
                        styleDefinitions[i],
                        classNameBits,
                        definitionBits,
                        length,
                    );
                } else if (this.isValidStyleDefinition(styleDefinitions[i])) {
                    classNameBits.push(styleDefinitions[i]._name);
                    definitionBits.push(styleDefinitions[i]._definition);
                    length += styleDefinitions[i]._len;
                } else {
                    throw new Error("Invalid Style Definition: Styles should be defined using the StyleSheet.create method.")
                }
            }
        }
        return length;
    }

    /**
     * Inject styles associated with the passed style definition objects, and return
     * an associated CSS class name.
     *
     * @param {boolean} useImportant If true, will append !important to generated
     *     CSS output. e.g. {color: red} -> "color: red !important".
     * @param {(Object|Object[])[]} styleDefinitions style definition objects, or
     *     arbitrarily nested arrays of them, as returned as properties of the
     *     return value of StyleSheet.create().
     */
    injectAndGetClassName(
        useImportant /* : boolean */,
        styleDefinitions /* : MaybeSheetDefinition[] */,
        selectorHandlers /* : SelectorHandler[] */
    ) /* : string */ {
        const classNameBits = [];
        const definitionBits = [];

        // Mutates classNameBits and definitionBits and returns a length which we
        // will append to the hash to decrease the chance of hash collisions.
        const length = this.processStyleDefinitions(
            styleDefinitions,
            classNameBits,
            definitionBits,
            0,
        );

        // Break if there aren't any valid styles.
        if (classNameBits.length === 0) {
            return "";
        }

        let className;
        if (process.env.NODE_ENV === 'production') {
            className = classNameBits.length === 1 ?
                `_${classNameBits[0]}` :
                `_${hashString(classNameBits.join())}${(length % 36).toString(36)}`;
        } else {
            className = classNameBits.join("-o_O-");
        }

        this.injectStyleOnce(
            className,
            `.${className}`,
            definitionBits,
            useImportant,
            selectorHandlers
        );

        return className;
    }
}